pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java

```diff
@@ -0,0 +1,495 @@
+            // have not been committed. This will ensure that they are.
+            if (startFromCheckpoint) {
+                commitTransactions(startFromCheckpointName);
+            }
+
+            // Schedule periodic task to initiate checkpoints.
+            // If any execution of this task takes longer than its period, then subsequent executions may start late, but will not concurrently execute.
+            initiateCheckpointExecutor.scheduleAtFixedRate(this::performCheckpoint, checkpointPeriodMs, checkpointPeriodMs, TimeUnit.MILLISECONDS);
+
+            // Start workers.
+            IntStream.range(0, numWorkers).forEach(workerIndex -> {
+                Worker worker = new Worker(workerIndex, scope, readerGroupName, startFromCheckpointName, outputStreamName, controllerURI);
+                workerExecutor.submit(worker);
+            });
+            try {
+                workerExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+            } catch (InterruptedException e) {
+            }
+        }
+
+        /**
+         * Initiate a checkpoint, wait for it to complete, and write the checkpoint to the state.
+         */
+        private void performCheckpoint() {
+            final String checkpointName = UUID.randomUUID().toString();
+            log.info("performCheckpoint: BEGIN: checkpointName={}", checkpointName);
+            try {
+                Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+                checkpointDirPath.toFile().mkdirs();
+
+                final Set<String> onlineReaders = readerGroup.getOnlineReaders();
+                log.info("performCheckpoint: onlineReaders ({})={}", onlineReaders.size(), onlineReaders);
+                log.info("performCheckpoint: Calling initiateCheckpoint; checkpointName={}", checkpointName);
+                CompletableFuture<Checkpoint> checkpointFuture = readerGroup.initiateCheckpoint(checkpointName, performCheckpointExecutor);
+                Checkpoint checkpoint = checkpointFuture.get(checkpointTimeoutMs, TimeUnit.MILLISECONDS);
+                // At this point, all workers have received and processed the checkpoint.
+                log.info("performCheckpoint: Checkpoint completed; checkpointName={}, positions={}", checkpointName, checkpoint.asImpl().getPositions());
+
+                Path checkpointPath = checkpointDirPath.resolve(PRAVEGA_CHECKPOINT_FILE_NAME);
+                log.info("Writing Pravega checkpoint to {}", checkpointPath);
+                try (FileOutputStream fos = new FileOutputStream(checkpointPath.toFile())) {
+                    fos.write(checkpoint.toBytes().array());
+                    fos.flush();
+                    fos.getFD().sync();
+                }
+
+                // Create "latest" file that indicates the latest checkpoint name.
+                // This file must be updated atomically.
```

Comment:
The checkpoint directory is created when the checkpoint is initiated. However it will not contain a complete checkpoint until the workers write their state and the master writes the stream cuts. The "latest" file is essentially a marker to indicate the latest checkpoint that completed. Should I create a "complete" marker file in the checkpoint directory to indicate this?